File: app/src/main/java/com/ddiehl/android/reversi/game/BoardSpace.java
package com.ddiehl.android.reversi.game;


import android.support.annotation.Nullable;

public class BoardSpace {

    private @Nullable ReversiColor color;
    private int x;
    private int y;

    public BoardSpace() {
        color = null;
    }

    public BoardSpace(int x, int y) {
        color = null;
        this.x = x;
        this.y = y;
    }

    public int x() {
        return x;
    }

    public int y() {
        return y;
    }

    protected BoardSpace copy() {
        BoardSpace copy = new BoardSpace(x, y);
        copy.color = color;
        return copy;
    }

    public boolean isOwned() {
        return color != null;
    }

    public @Nullable ReversiColor getColor() {
        return color;
    }

    public void setColor(@Nullable ReversiColor c) {
        color = c;
    }

    public ReversiColor flipColor() {
        if (color == null) {
            return null;
        } else {
            color = (color == ReversiColor.Dark) ? ReversiColor.Light : ReversiColor.Dark;
            return color;
        }
    }
}

File: app/src/main/java/com/ddiehl/android/reversi/game/BoardSpace.kt
package com.ddiehl.android.reversi.game

class BoardSpace(private val row: Int, private val col: Int) {

    var color: ReversiColor? = null

    fun x(): Int {
        return col
    }

    fun y(): Int {
        return row
    }

    fun copy(): BoardSpace {
        val copy = BoardSpace(row, col)
        copy.color = color
        return copy
    }

    val isOwned: Boolean
        get() = color != null

    fun flipColor(): ReversiColor? {
        if (color == null) {
            return null
        } else {
            color = if (color == ReversiColor.Dark) ReversiColor.Light else ReversiColor.Dark
            return color
        }
    }
}
